### PR TITLE
Crossword

### LINKED ISSUE #15: Crossword
We want to create a crossword puzzle which uses the words from a piece of text for the words and the definitions for the hints.
The instructor will be able to choose which definitions to use for each word that is in the crossword.

We are currently doing this crossword as a start for something bigger. We want to able to do something with the bigger context of the words in the sentence.

To-dos:
- [x] Backend functions to generate crossword grid based on words and pair it with definitions
- [x] Wireframe on Figma
- [x] Decide which content the crossword will be solved from.
    - [x] Definition: boring and simple, think again
    - [ ] Parts of Speech: easy but not contextual enough
    - [ ] Reading Comprehension questions: Very Contextual but painful to set up (maybe even not possible)
    - [x] Fill in the blank Sentence with word list added: pust words in context
- [x] Build front end interface:
    - [x] Crossword interface for filling in answers
    - [x] Definitions / hints
    - [x] A way to check your answer

### PR DESCRIPTION
**Pull request by sharabhuiyan**
Created a crossword game:

-  Generate a crossword given input text.
- Created a crossword generation algorithm that finds the first valid crossword
- Each crossword that is generated is different because we randomize the order of the words inputted 
- When a word is found, the text in a row/column turn green
- Crossword contains 3 modes (easy, medium, hard)
    - Easy mode:  definitions and clues are linked and crossed out when word is found
    -  Medium mode: definitions and clues are not linked together, only clues are crossed off when word is found
    - Hard mode: there are extra definitions to throw students off 
    
image
image
image

### CODE REVIEW

File: backend/app/analysis/crosswords.py [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+    :param all_words: datd structure containing all words to put into crossword
+    :param max_num_words: number of words that can be made into crossword
+    :return: list of words to be used (in order of how they will be placed in the grid)
+    """
+    words = set()
+    while len(words) < max_num_words:
+        indx = random.randint(0, len(all_words) - 1)
+        while all_words[indx] not in words:
+            words.add(all_words[indx])
+    return list(words)
```
Comment (MBJean) [inline backend/app/analysis/crosswords.py]: This function looks like it does the trick, but two things stand out to me:
1. I think that supplying a `max_num_words` that is greater than or equal to `len(all_words)` will cause an endless loop. Could we guard against that possibility?
2. Regardless of the above, the `random.randint` solution here could create duplicate attempts to insert the same word into `words` and duplicate the lookup of `not in`. This seems unlikely to cause a performance problem given the maximum likely length of `all_words`, but since we're already relying on `random` here, I wonder if we could shortcut some of this manual work by using `random.shuffle` as we do in functions below. i.e., something like:

```
def rand_words(all_words, max_num_words):
  output = all_words.copy()
  random.shuffle(output)
  return output[0:max_num_words]
```

File: backend/app/views.py [hunk 2]
```diff
@@ -130,7 +134,7 @@ def update_text(request):
     API endpoint for updating title, content, and modules for a given piece of text given
     the id of the text.
     """
-    body = json.loads(request.body.decode('utf-8'))
+    body = json.loads(json.dumps(request.data))
```
Comment (MBJean) [inline backend/app/views.py]: If you have a second, could you write out a few words explaining why this change is necessary, and why we need to both `dumps` and `loads` here and on line 123?

File: backend/app/analysis/crosswords.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        else:
+            words = all_words.copy()
+        words.sort(key=len, reverse=True)
+        grid_size = 15
+        grid = []
+        for i in range(grid_size):
+            row = []
+            for j in range(grid_size):
+                row.append(0)
+            grid.append(row)
```
Comment (MBJean) [inline backend/app/analysis/crosswords.py]: In this function, `grid_size` and `grid` are only used to pass into `make_crossword`, i.e., the assignment of these variables seems relevant to the making of the crossword (what's called in line 282) not the getting of it (the current function). Do you think it would better encapsulate functionality to move the `grid_size` and `grid` assignments/creation into `make_crossword`? Doing so would allow us to get rid of `grid` as an argument for `make_crossword`. This makes some additional sense to me: the `grid` at creation time is simply a product of the `grid_size`, so providing both `grid` and `grid_size` as arguments seems somewhat duplicative.

File: backend/app/analysis/crosswords.py [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+
+def make_crossword(grid, word_list, grid_size):
+    """
+    Returns a grid with all the words in a crossword format and their corresponding
+    clues.
+    :param: grid (list): 2D list representing the format of the crossword
+            word_list (list): contains words to be placed into crossword
+            grid_size (int): represents dimension of grid (grid is a square)
+    """
+    first_word = word_list[0]
```
Comment (MBJean) [inline backend/app/analysis/crosswords.py:165]: What is the consequence of not sorting the `word_list` by reverse `len(word)`? We perform that sort outside `make_crossword`, so there's the possibility that future us could attempt to use this function again without remembering that it depends on that sort. Would it be worth adding a note in the docstring, or perhaps even performing the sort in this function?

File: backend/app/analysis/crosswords.py [hunk 5]
```diff
@@ -0,0 +1,306 @@
...
+    return True
+
+
+def is_within_bounds(word_len, row, column, direction, grid_dim):
+    """ Returns whether the given word is withing the bounds of the grid.
+    """
+    # checks if word is in within grid boundaries
+    return ((direction == "across" and column + word_len <= grid_dim[0] - 1)
+            or (direction == "down" and row + word_len <= grid_dim[1] - 1)) and (
+               row >= 0 and column >= 0)
```
Comment (MBJean) [inline backend/app/analysis/crosswords.py:58]: Since `make_crossword` assumes/requires the grid to be a square, could we simplify `grid_dim` as simply `grid_size`?

File: frontend/src/crosswordView/crosswordView.js [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    }
+
+    updateCell = (e, row, col) => {
+        const grid = this.state.grid;
+        const curLetter = e.target.value.toUpperCase();
+        if (ALPHA.includes(curLetter)) {
+            grid[row][col] = curLetter;
+            // Check if a word has been found
+            this.setState({ grid }, this.updateFound);
+        }
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js:101]: I'd recommend a refactor of this method. If I read this correctly, we've copied a reference to `this.state.grid` and assigned it to `grid`, but haven't deep copied the value of `this.state.grid` (the 2D array) itself. The consequence is that `grid[row][col] = curLetter` is actually modifying `state.grid[row][col]` without using React's `setState` method. We're then using `setState` to set `state.grid` to itself (since `grid` is just a reference to `state.grid`), which forces the rerender we want to update the view. This presumably works as expected, although we've technically modified the data structure (the 2D array `grid`) twice, only one modification of which will trigger the React rerender. It's been my experience that directly modifying state without using the React `setState` [can cause really ambiguous bugs](https://daveceddia.com/why-not-modify-react-state-directly/) (although admittedly in this case it seems relatively harmless). I think a more React-specific solution would fully swap out the grid for a copy that has had the appropriate 'cell' modified. Something like:

```
const new_grid = [];
const curLetter = e.target.value.toUpperCase();

for (let new_grid_row = 0; new_grid_row < this.state.grid.length; new_grid_row++) {
  const holder = [];
  for (let new_grid_column = 0; new_grid_column < this.state.grid[new_grid_row].length; new_grid_column++) {
    holder.push(this.state.grid[new_grid_row][new_grid_column]);
  }
  new_grid.push(holder);
}

if (ALPHA.includes(curLetter)) {
  new_grid[row][col] = curLetter;
  this.setState({ grid: new_grid }, this.updateFound)
}
```

Or a more functional approach:

```
this.setState(
  (state) => ({
    grid: state.grid.map(
      (temp_row, temp_row_index) => temp_row.map(
        (temp_col, temp_col_index) => row === temp_row_index && col === temp_col_index ? curLetter: temp_col
      )
    )
  }),
  this.updateFound
);
```

One more note: the second code snippet above makes use of the `updater` function pattern described in the [React docs](https://reactjs.org/docs/react-component.html#setstate). The justification for the pattern is outlined well [here](https://medium.com/@wisecobbler/using-a-function-in-setstate-instead-of-an-object-1f5cfd6e55d1). Regardless, the above two code snippets should achieve the same effect.

File: frontend/src/crosswordView/crosswordView.js [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+
+    componentDidMount = async () => {
+        try {
+            const apiURL = `/api/get_crossword/${this.props.textID}/${this.props.partOfSpeech}`;
+            const response = await fetch(apiURL);
+            const data = await response.json();
+            const emptyGrid = data.solution.map((row) => row
+                .map((cell) => (ALPHA.includes(cell) ? '' : '#')));
+            this.setState({
+                crosswordData: data,
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js]: Did you all consider flattening `crosswordData` into its properties before setting it into the component state? That is: `state.crosswordData` has an unspecified set of properties (I can see from the code below that those properties include `solution`, `clues`, and `definitions`). It looks like the only way to know what those properties are is to consult the API directly to determine the shape of the data. I wonder if it'd be preferable to set expectations around the shape of the return data from that endpoint here, as well. 

We could destructure `response.json()` as:

```
const { solution, clues, definitions } = response.json();
```

So, instead of:

```
state: {
  ...,
  crosswordData: {
    solution: ...,
    clues: ...,
    definitions: ...
   }
}
```

we simply set:

```
state: {
  ...,
  solution: ...,
  clues: ...,
  definitions: ...
}
```

File: frontend/src/crosswordView/crosswordView.js [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+                    ))}
+                </ul>
+            );
+        default:
+            return '';
+        }
+    }
+
+    modalHandler = () => {
+        this.setState({ showModal: !this.state.showModal });
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js]: Please see my comment above about the `updater` function pattern in React's `setState`, and some of the challenges that might occur when referencing state directly as the value of a `setState` update statement.

File: frontend/src/crosswordView/crosswordView.js [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+    }
+
+    render() {
+        if (!this.state.crosswordData) {
+            return (<LoadingPage />);
+        }
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js]: While this works, a safer and more explicit pattern pattern would be to introduce a `loading` property to the component's state and key this UI element off of it.

File: frontend/src/crosswordView/crosswordView.js [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+                            >
+                                {k + 1}) {found[k].down
+                                    ? clue.down.clue
+                                    : coverWord(clue.down.word, clue.down.clue)
+                                }
+                            </div>
+                        );
+                    }
+                    return '';
+                })
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js]: I can see this pattern used in a couple of places (i.e., map over an array, guard against the existence of a property, return `''` or `null` as the fallback). We could possibly use the [Array.filter method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) to ensure the array only contains meaningful data and thereby reduce the amount of boilerplate around this pattern.

File: frontend/src/crosswordView/crosswordView.js [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+                        : <>
+                            {instructionModal}
+                            <h1 className='crossword-title'>
+                                Crossword
+                                <button className="btn btn-outline-light btn-circle mx-3"
+                                    style= {{ 'border': '3px solid', 'fontSize': '20px' }}
+                                    onClick={this.modalHandler}>
+                                    <b>?</b>
+                                </button>
+                            </h1>
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js]: I can see an opportunity to improve the [semantics](https://developer.mozilla.org/en-US/docs/Glossary/Semantics) of this portion of the UI. 

Generally, for the sake of accessibility (particularly screen readers) and of [SEO](https://developer.mozilla.org/en-US/docs/Glossary/SEO), a single rendered HTML page should have a fixed `<h1>` element that [does not change](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements). 

The pattern I see in this code is to swap out the `<h1>` and its content depending on the user's engagement with the difficulty selection buttons. A more conventional pattern might be to render a single, fixed `<h1>` element (with the content being, say, 'Crossword'), and have a `<form>` UI component that contains both a `<label>` describing the user's choices and `<input>` or `<button>` selection options. 

There'd be a whole host of other considerations that would follow from that change (one being: how to announce to a screen reader that there's new content on the page when a user interactions with the difficulty buttons), but we can tackle accessibility more broadly in future PRs. In this case, I think it'd be worth reworking how this `<h1>` element is being modified to take us one step closer toward more semantic HTML.

For those interested in web software development, writing semantic HTML is a foundational skill, so you may find value in skimming through the MDN docs I linked above.

File: backend/app/views.py [hunk 12]
```diff
@@ -155,6 +159,43 @@ def delete_text(request):
...
+
+    # Add all definitions from input text for hard difficulty
+    crossword_data['definitions'] = []
+    for defs in definitions.values():
+        pos_def = defs.get(part_of_speech, None)
+        if pos_def:
+            crossword_data['definitions'].append(pos_def[0])
+
+    random.shuffle(crossword_data['definitions'])
+    return Response(crossword_data)
```
Comment (MBJean) [inline backend/app/views.py:178]: Would it be possible to refactor this and push lines 179-195 down into `get_crosswords`? That way we could keep the API layer's responsibilities limited (as we do below with `get_quiz_data`) and more easily unit test the various pieces.

File: frontend/src/crosswordView/crosswordView.js [hunk 13]
```diff
@@ -0,0 +1,343 @@
...
+                for (let r = 0; r < clue.down.word.length; r++) {
+                    if (grid[clue.row + r][clue.col] !== clue.down.word.charAt(r)) {
+                        correct = false;
+                        break;
+                    }
+                }
+                found[k].down = correct;
+            }
+        });
+        this.setState({ found });
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js]: Please see my comment above for the `updateCell` method. In this method, I can see that we're again modifying state twice: once without React's `setState` and once with it.

File: frontend/src/crosswordView/crossword.js [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+    }
+}
+Crossword.propTypes = {
+    grid: PropTypes.array,
+    updateCell: PropTypes.func,
+    getClueNumber: PropTypes.func,
+    isCorrect: PropTypes.func,
+};
+
+export default Crossword;
```
Comment (MBJean) [inline frontend/src/crosswordView/crossword.js:114]: I think this component is very well constructed, great work! I appreciate the consistency in the use of variables names throughout this component, though I'd encourage everyone not to be afraid to expand on variable names for the sake of readability. It looks like `r` and `c` in this component are consistently used to refer to something like `row_index` and `column_index`, respectively. It'd probably help our future selves out to expand those variable names to make it clear in everyone context what they refer to. One example: there's a little inconsistency there when we pass up those index references to `this.props.updateCell` in the parent component, which refers to them as `row` and `col`, respectively. I think it'd be worth it to take a quick pass at both this and `crosswordView` to ensure that we're consistent in how we refer to these values.

File: frontend/src/crosswordView/crosswordView.scss [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+.give-up {
+    margin: 10px 10px 10px 0;
+}
+
+.difficulty-button {
+    height: 100px;
+    width: 200px;
+    font-size: 40px;
+
+}
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.scss:34]: The most nitpick of nitpicks: it looks like there's an errant empty line here.

File: frontend/src/crosswordView/crosswordView.scss [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    background: rgba(105, 64, 238, 0.25);
+    border-radius: 10px;
+    padding-bottom: 20px;
+    max-height: 300px;
+    overflow-y: scroll;
+}
+
+.definition-title {
+
+}
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.scss]: If we don't intend to make use of this className, let's get rid of it.

File: frontend/src/crosswordView/crosswordView.scss [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+    flex-direction: column;
+}
+
+.crossword-row {
+    display: flex;
+    justify-content: center;
+}
+
+$cell-width: 50px;
+$cell-height: 50px;
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.scss]: This is a matter of personal style, probably, but I think it's a good idea to pull up SCSS variable assignments to the top of the file so that it's easy for our future selves to find them.

File: frontend/src/crosswordView/crosswordView.scss [hunk 18]
```diff
@@ -0,0 +1,121 @@
...
+    height: $cell-height;
+    width: $cell-width;
+    position: relative;
+    border: none;
+    cursor: default;
+
+    &:focus {
+        outline: none;
+    }
+}
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.scss:136]: This CSS looks well formed! There's a few topics that might be worth discussing, especially if any one has a particular interest in presentation. 

[Using relative units for sizing](https://www.24a11y.com/2019/pixels-vs-relative-units-in-css-why-its-still-a-big-deal/). This is a big aid in responsive design and worth thinking about, if you haven't already.

[Alphabetical ordering for CSS properties](https://medium.com/@jerrylowm/alphabetize-your-css-properties-for-crying-out-loud-780eb1852153). This is a big aid in avoiding headaches when you inevitably need to update old CSS.

Since I'm coming into this well into the project, please feel free to ignore if you all have common patterns that you're following already.

File: frontend/src/crosswordView/crosswordView.js [hunk 19]
```diff
@@ -0,0 +1,343 @@
...
+                    ? <div className="backdrop" onClick={this.modalHandler}>
+                    </div>
+                    : null
+            }
+            <div className="Modal modal-content" style={{
+                transform: showModal ? 'translateY(0)' : 'translateY(-100vh)',
+                opacity: showModal ? 1 : 0,
+            }}>
+                <div className="modal-header">
+                    <h5 className="modal-title">Instructions</h5>
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js]: Please see my other comment about semantic HTML. The use of the `<h5>` here violates the [guidelines around HTML headers](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements#usage_notes) since it doesn't follow in the appropriate header sequence (i.e., sequentially, `<h1>`, `<h2>`, etc.).

File: frontend/src/crosswordView/crosswordView.js [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+                                    </div>
+                                </div>
+                            </div>
+                        </>
+                    }
+
+                </div>
+                <Footer/>
+            </>
+        );
```
Comment (MBJean) [inline frontend/src/crosswordView/crosswordView.js:345]: There's a lot going on in this `render` method, which is making it a bit challenging to grok fully and to skim for things like semantics and accessibility concerns. [This article](https://medium.com/front-end-weekly/react-quick-tip-split-up-your-render-method-5eb30e432e2) outlines a few techniques for keeping your render functions slim, with the general idea being to follow the basic "single-responsibility principle" of programming.

**Review by MBJean — CHANGES_REQUESTED** [review]
Hi, everyone! I'm picking up the review for this PR. 

Overall, this is fantastic work. There's real a sophistication to how you all have kept each part of the stack responsible for its own concerns. 

I've left a lot of comments below on various places that I can see for improvement or refactor. The goal of most of the suggested refactor is broadly the same: to help our future selves out when we inevitably need to return to this code and continue working on it.

With all of my reviews, please always feel comfortable pushing back or explaining your reasoning. If anyone has any questions or needs clarification, don't hesitate to reach out and we can meet in office hours, the Friday hack session, or schedule some specific time to meet.

One thing that I will consider a blocker: I'd love to see some testing around this work.

And one thing I'm interested in your opinions on: how challenging do you all think it'd be to avoid clues lining up in such a way that they create non-clued, nonsense words (like IPEO and LKSC in the screenshot below)?

Screen Shot 2021-03-24 at 11 43 33 AM

File: backend/app/views.py (hunk 2 above)
Comment (JasonLin43212) [reply backend/app/views.py]: The main motivation for the change was that at the time, we were getting a strange error about not being able to read from the `request.body` twice and when I searched up, it was related to Django middleware, which was really confusing.

However, I tested the code using `json.loads(request.body.decode('utf-8'))` now and it seems fine. Should I change it back?

File: backend/app/analysis/crosswords.py (hunk 1 above)
Comment (sharabhuiyan) [reply backend/app/analysis/crosswords.py]: Thank you for the feedback! 
1. In the `get_crossword` function, we do check to see if the` len(all_words)< max_num_words` in this block of code. 
image
2.  You brought up a good point! I have made the changes necessary.

File: backend/app/analysis/crosswords.py (hunk 4 above)
Comment (sharabhuiyan) [reply backend/app/analysis/crosswords.py:165]: We sorted the `word_list` by `len(word)` because it is easier to place the hardest (longest) word into the crossword because it allows us to have more possibilities for connecting words in the future. I have changed the code so that we perform the sort within `make_crossword`.

File: backend/app/analysis/crosswords.py (hunk 3 above)
Comment (sharabhuiyan) [reply backend/app/analysis/crosswords.py]: I agree with this statement. I have made the changes, thanks!

File: backend/app/analysis/crosswords.py (hunk 5 above)
Comment (sharabhuiyan) [reply backend/app/analysis/crosswords.py:58]: We decided to use `grid_dim`  in case we were to edit the code in the future and allow other grid shapes.

File: backend/app/views.py (hunk 12 above)
Comment (taokinbo) [reply backend/app/views.py:178]: Definition and example setting has now been moved into the get_crosswords functions.

**Comment by taokinbo** [comment]
@MBJean We have made all relevant changes that were requested and gave reasons for specific changes that were not made under your review comments. We have also looked into the nonsense words that were previously being made and changed our validation code to no longer allow them: 
image

We eagerly await your rereview :)

File: backend/app/analysis/crosswords.py (hunk 5 above)
Comment (MBJean) [reply backend/app/analysis/crosswords.py:58]: 👍

**Comment by sharabhuiyan** [comment]
> @MBJean We have made all relevant changes that were requested and gave reasons for specific changes that were not made under your review comments. We have also looked into the nonsense words that were previously being made and changed our validation code to no longer allow them:
> image
> 
> We eagerly await your rereview :)

We added tests for the functions that are essential to creating crosswords! Thank you :)

**Review by MBJean — APPROVED** [review]
This looks great! Great job figuring out how to meaningfully test the crossword creation. Merging.
